app/models/concerns/workflow_extraction.rb

```diff
@@ -80,22 +80,59 @@ def populate_ro_crate(crate)
     rescue WorkflowDiagram::UnsupportedFormat
     end
 
-    authors = creators.map { |person| crate.add_person(nil, person.ro_crate_metadata) }
-    others = other_creators&.split(',')&.collect(&:strip)&.compact || []
-    authors += others.map.with_index { |name, i| crate.add_person("creator-#{i + 1}", name: name) }
-    crate.author = authors
-    crate['provider'] = projects.map { |project| crate.add_organization(nil, project.ro_crate_metadata).reference }
-    crate.license = license
     crate.identifier = ro_crate_identifier
     crate.url = ro_crate_url('ro_crate')
-    crate['isBasedOn'] = source_link_url if source_link_url
-    crate['sdPublisher'] = crate.add_person(nil, contributor.ro_crate_metadata).reference
-    crate['sdDatePublished'] = Time.now
+
+    workflow = is_a_version? ? self.parent : self
+    merge_entities(crate, workflow) if workflow
+
+    crate['isBasedOn'] = source_link_url if source_link_url && !crate['isBasedOn']
+    crate['sdDatePublished'] = Time.now unless crate['sdDatePublished']
     crate['creativeWorkStatus'] = I18n.t("maturity_level.#{maturity_level}") if maturity_level
+#    authors = creators.map { |person| crate.add_person(nil, person.ro_crate_metadata) }
+#    others = other_creators&.split(',')&.collect(&:strip)&.compact || []
+#    authors += others.map.with_index { |name, i| crate.add_person("creator-#{i + 1}", name: name) }
+#    crate.author = authors
+#    crate['provider'] = projects.map { |project| crate.add_organization(nil, project.ro_crate_metadata).reference }
+#    crate.license = license
+#    crate['sdPublisher'] = crate.add_person(nil, contributor.ro_crate_metadata).reference
 
     crate.preview.template = PREVIEW_TEMPLATE
+    # brute force deletion as I cannot track down where it comes from
+    crate.contextual_entities.delete_if { |c| c['@id'] == '#ro-crate-preview.html' }
   end
 
+  def merge_entities (crate, workflow)
+    workflow_struct = Seek::BioSchema::Serializer.new(workflow).json_representation
+
+    context = {
+      '@vocab' => 'https://somewhere.com/'
+    }
+    workflow_struct['@context'] = context
+    workflow_struct.except!('encodingFormat')
+
+    flattened = JSON::LD::API.flatten(workflow_struct, context)
```

Comment:
how long does this take?


In my experience, using any kind of linked-data aware library takes 10+ seconds to do any operation, and in SEEK we can check a resource's RO-Crate several times in a request.

Reply:
It could be hand coded I guess. Would that be better?

Reply:
possibly, and would make the above hack unnecessary

Reply:
For an initial version - I do not think this is necessary